embassy-stm32/src/eth/mod.rs

```diff
@@ -30,13 +30,42 @@ pub struct PacketQueue<const TX: usize, const RX: usize> {
 
 impl<const TX: usize, const RX: usize> PacketQueue<TX, RX> {
     pub const fn new() -> Self {
-        const NEW_TDES: TDes = TDes::new();
-        const NEW_RDES: RDes = RDes::new();
-        Self {
-            tx_desc: [NEW_TDES; TX],
-            rx_desc: [NEW_RDES; RX],
-            tx_buf: [Packet([0; TX_BUFFER_SIZE]); TX],
-            rx_buf: [Packet([0; RX_BUFFER_SIZE]); RX],
+        #[cfg(feature = "nightly")]
+        {
+            let mut this = core::mem::MaybeUninit::uninit();
+            unsafe {
+                Self::init(&mut this);
+                this.assume_init()
+            }
+        }
+        #[cfg(not(feature = "nightly"))]
+        {
+            const NEW_TDES: TDes = TDes::new();
+            const NEW_RDES: RDes = RDes::new();
+            Self {
+                tx_desc: [NEW_TDES; TX],
+                rx_desc: [NEW_RDES; RX],
+                tx_buf: [Packet([0; TX_BUFFER_SIZE]); TX],
+                rx_buf: [Packet([0; RX_BUFFER_SIZE]); RX],
+            }
+        }
+    }
+
+    // Allow to initialize a Self without requiring it to go on the stack
+    #[cfg(feature = "nightly")]
+    pub const unsafe fn init(this: &mut core::mem::MaybeUninit<Self>) {
+        let this: &mut Self = unsafe { this.assume_init_mut() };
```

Comment:
creating a `&mut` to uninitialized data is undefined behavior. You have to do all writes to initialize using raw pointers, only after that you can create the `&mut`.